src/config/mongo.rs

```diff
@@ -0,0 +1,597 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    path::PathBuf,
+    time::Duration,
+};
+
+use cfg_if::cfg_if;
+use mongodb::{
+    bson::Document,
+    options::{
+        ClientOptions,
+        ReadConcern,
+        ReadPreferenceOptions,
+        ServerApi,
+        WriteConcern,
+    },
+};
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+/// A clone of the `ClientOptions` structure from the `mongodb` crate
+/// which can be Serialized.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct MongoConfig {
```

Comment:
Should all of these options be configurable to the user? If not I think we should go with the builder pattern here. The builder can then be serializable and contains all the options that the user can specify (like address and such). In the `finish`/`build` step of the builder we can set the rest and return a `mongodb::ClientOptions` object.

If we follow this pattern for all major components, we can then aggregate all these builders to obtain our config file format.

Reply:
The struct was originally designed to be read from a file, so I didn't bother with a builder, but yes it could use one. How do you feel about `derive-builder`?

Reply:
Nevermind, it's not suitable for this case. I added some `with_` methods instead.

Reply:
Looking good! We are probably going to need some of these configuration options in the future, but to make reviewing simpler for everyone, I think I would only add the ones we actually use for now. When can then add additional parameters to config as we need them.